lib/recursive-methods.rb

```diff
@@ -1,49 +1,140 @@
 # Authoring recursive algorithms. Add comments including time and space complexity for each method.
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity:  O(n)  
+# Space complexity: O(1)
+
 def factorial(n)
-    raise NotImplementedError, "Method not implemented"
+    # if wrong argument is provided
+    if n < 0
+        raise ArgumentError
+    end
+    #base case
+    if n == 0
+        return 1
+    else
+        return n * factorial(n-1)
+    end
 end
 
-# Time complexity: ?
-# Space complexity: ?
+
+# Time complexity: O(n)
+# Space complexity: O(n)
 def reverse(s)
-    raise NotImplementedError, "Method not implemented"
+    #base case
+    if s.length == 0
+        return ""
+    else
+        return s[-1] + reverse(s[0..-2])
+    end
+        
 end
 
-# Time complexity: ?
-# Space complexity: ?
+
+# Time complexity: O(n) 
+# Space complexity: O(n^2)
 def reverse_inplace(s)
```

Comment:
The advantage of doing this in place is that you're not creating a new array with each recursive call.  So the space complexity is O(n), just the cost of the system stack.